include/oneapi/internal/execution_context_sycl.h

```diff
@@ -42,166 +42,175 @@ namespace interface1
+            if (status != nullptr && !status->ok())
+                return;
+            if (!res)
+            {
+                size_t id = get_id(name, status);
+                if (status != nullptr && !status->ok())
+                    return;
+                arr[id] = new Entry();
+                arr[id]->name = name;
+                arr[id]->obj = obj;
+                return;
+            }
+            setStatus(status);
         }
 
-        const char* getName()
+        services::SharedPtr<T> get(const char *name, services::Status * status = nullptr)
         {
-            return _name.c_str();
+            size_t id = get_id(name, status);
+            if (status != nullptr && !status->ok())
+                return services::SharedPtr<T>();
+
+            if (arr[id] != nullptr)
+                return arr[id]->obj;
+
+            setStatus(status);
+            return services::SharedPtr<T>();
         }
     };
+    static const size_t SIZE_MAP_PROGRAM = 1024;
+    static const size_t SIZE_MAP_KERNEL  = 4096;
+    Map <OpenClProgramRef, SIZE_MAP_PROGRAM> programMap;
+    Map <KernelIface, SIZE_MAP_KERNEL> kernelMap;
+
+    OpenClProgramRef * _clProgramRef;
+
+    ExecutionTargetId _executionTarget;
+    cl::sycl::queue & _deviceQueue;
 
 public:
-    explicit OpenClKernelFactory(cl::sycl::queue & deviceQueue) : _clProgramRef(nullptr), _executionTarget(ExecutionTargetIds::unspecified), _deviceQueue(deviceQueue)
-    {}
+    explicit OpenClKernelFactory(cl::sycl::queue & deviceQueue) : _clProgramRef(nullptr),
+        _executionTarget(ExecutionTargetIds::unspecified), _deviceQueue(deviceQueue) {}
 
-    void build(ExecutionTargetId target, const char * key, const char * program, const char * options = "",
+    ~OpenClKernelFactory() DAAL_C11_OVERRIDE {}
+
+
+    void build(ExecutionTargetId target, const char * name, const char * program, const char * options = "",
                services::Status * status = nullptr) DAAL_C11_OVERRIDE
     {
         // TODO: Thread safe?
-        // TODO Rework of "cache"
-
-        uint64_t id = hash(key) % SIZE_CACHE_PROGRAM;
 
-        while (_clProgramCache[id].getProgram() != nullptr && strcmp(_clProgramCache[id].getName(), key) != 0)
+        const bool res = programMap.contain(name, status);
+        if (status != nullptr && !status->ok())
+            return;
+        if (!res)
         {
-            id = (id + 1) % SIZE_CACHE_PROGRAM;
-        }
-
-        if (_clProgramCache[id].getProgram())
-        {
-            _clProgramRef    = _clProgramCache[id].getProgram();
-            _executionTarget = target;
-        }
-        else
-        {
-            _clProgramCache[id].setProgram(new OpenClProgramRef(_deviceQueue.get_context().get(),
-                                           _deviceQueue.get_device().get(), key, program, options, status));
+            programMap.add(name, services::SharedPtr<OpenClProgramRef>(new OpenClProgramRef(_deviceQueue.get_context().get(),
```

Comment:
It is better to make a creation of `services::SharedPtr<OpenClProgramRef>` in a separate line